### PR TITLE
63 verify moab structure

### PR DESCRIPTION
**Pull request by tallenaz**
Adds a method and specs for checking the structure of a moab, and returning information on the structure for the caller to be able to log.

TODO: Rebase off master

Thoughts (sarav):
- Tony and I didn't know where the best place to keep this method was, so we just decided to put it where we put the current_version and object_size methods. If there is somewhere else that is better to put it please let us know. I thought it was good to put it in the storage_repository and also the storage_service so when you call the method you can call it by doing ``` Stanford::StorageServices.verify_no_nested_moabs('spec/fixtures/bad_root01/bad_moab_storage_trunk/xx/000/xx/0000/xx000xx0000') ```

- Used John's method of sending a response_code so then the PC Rails APP will have some sort of 'legend' that will take in those response_codes and put out the correct Rails.log ?? Seemed like this was a good first pass. (If anyone has another way to do this please let me know!)

- Lastly, there are some edge cases that we did not account for, and was wondering if it is worth to account. When a version directory exists and only a file is under it, like a text.txt file (V0001/extra_file.txt) , the program will through an error - is this worth adding to the test cases?
 -- The cases accounted for are shown in the specs

connects #63

### CODE REVIEW

**Comment by ndushay** [comment]
FYI:  it was John's idea to do the return code => message results for PreservedObjectHandler.

**Comment by ndushay** [comment]
I would move this to StorageObject.  it makes more sense there, no?  It's really about a particular object.  I wonder if we should even keep the other methods in StorageRepository -- what do you think?

File: lib/moab/storage_repository.rb [hunk 1]
```diff
@@ -115,19 +127,37 @@ def object_size(object_id, include_deposit=false)
...
+    # @return [Array] of hashes which contain messages for the caller
+    def verify_no_nested_moabs(path)
+      version_directories = list_sub_dirs(path).sort # sort for travis
+      results = []
+      version_directories.each do |version|
+        version_path = path + "/#{version}"
+        version_sub_dirs = list_sub_dirs(version_path).sort
+        version_sub_dir_count = version_sub_dirs.count
+        check_sub_dirs(version_sub_dir_count, version_sub_dirs, results)
+        data_dir_path = version_path + "/#{version_sub_dirs[0]}"
```
Comment (tallenaz) [inline lib/moab/storage_repository.rb]: Here, and elsewhere, we could use string interpolation, so `"#{version_path}/#{version_sub_dirs[0]}"` etc.

File: lib/moab/storage_repository.rb [hunk 2]
```diff
@@ -139,6 +169,57 @@ def store_new_object_version(druid, bag_pathname )
...
+
+    def result_hash(response_code, addl=nil)
+      { response_code => result_code_msg(response_code, addl) }
+    end
+
+    def result_code_msg(response_code, addl=nil)
+      format(RESPONSE_CODE_TO_MESSAGES[response_code], addl: addl)
+    end
+
+    def sub_dir(dir=nil)
```
Comment (ndushay) [inline lib/moab/storage_repository.rb]: I would sooner have EXPECTED_DATA_SUBDIRS = ["content", "metadata"].freeze  and EXPECTED_VERSION_SUBDIRS = ["data", "manifests"].freeze  than a method like this.  I think maybe it would be better for now to split/repeat the code currently in check_sub_dirs into  check_version_dir_content and check_data_dir_content and check_metadata_dir_content  ... this is feeling a little hard to follow / a little too clever - hard to make a change for one case without impacting the other case.

File: spec/unit_tests/moab/storage_repository_spec.rb [hunk 3]
```diff
@@ -76,6 +76,61 @@
     expect(storage_repo.object_size('jq937jp0017')).to be_between(345_000, 346_000)
   end
 
+  context '#verify_no_nested_moabs' do
+    let(:path) { 'spec/fixtures/bad_root01/bad_moab_storage_trunk/xx/000/xx/0000/xx000xx0000' }
+    let(:verification_array) { storage_repo.verify_no_nested_moabs(path) }
+
+    it 'returns correct data structure' do
+      expect(storage_repo.verify_no_nested_moabs(path)).to be_kind_of Array
```
Comment (ndushay) [inline spec/unit_tests/moab/storage_repository_spec.rb]: maybe show that an item in the array is the expected structure, also?  That all array elements are the expected structure (allowed keys, etc.)

File: lib/moab/storage_repository.rb [hunk 4]
```diff
@@ -115,19 +127,37 @@ def object_size(object_id, include_deposit=false)
       size
     end
 
+    # @param path [String] The identifier of the path to a digital object
+    # @return [Array] of hashes which contain messages for the caller
+    def verify_no_nested_moabs(path)
+      version_directories = list_sub_dirs(path).sort # sort for travis
```
[the remark this answers is not in the document]
Comment (SaravShah) [reply lib/moab/storage_repository.rb]: yes - I'm working on figuring out that logic and once I do i think it will fit in this method

File: lib/moab/storage_repository.rb [hunk 5]
```diff
@@ -115,19 +127,37 @@ def object_size(object_id, include_deposit=false)
...
+      results = []
+      version_directories.each do |version|
+        version_path = path + "/#{version}"
```
[the remark this answers is not in the document]
Comment (SaravShah) [reply lib/moab/storage_repository.rb]: To the first comment:  I think when it does that when we use the ``` check_sub_dirs ``` method and return ```expected_dirs ``` method which returns the `CORRECT_DIR` response code? Is that what you mean? Like if the version directory contained ONLY the "data" and "manifest" directory we would get the `CORRECT_DIR` response code.  Not sure if i'm interpreting your question correctly

[the remark this answers is not in the document]
Comment (SaravShah) [reply lib/moab/storage_repository.rb]: To the second comment: Yes. I believe tony is working on that logic and it should be implemented in here! :)

[the remark this answers is not in the document]
Comment (ndushay) [reply lib/moab/storage_repository.rb]: yes - you are right, sarav.  I was reading the code top down in the github GUI.  But maybe this also points out that it's a bit less than obvious?

File: lib/moab/storage_repository.rb [hunk 6]
```diff
@@ -139,6 +169,57 @@ def store_new_object_version(druid, bag_pathname )
       new_version
     end
 
-  end
+    private
+
+    def check_sub_dirs(sub_dir_count, sub_dirs, results, dir=nil)
+    # assuming case statements are more expensive than if-elsif
```
[the remark this answers is not in the document]
Comment (SaravShah) [reply lib/moab/storage_repository.rb]: I did a quick search and found  if-elsif is faster than case statements

File: lib/moab/storage_repository.rb [hunk 7]
```diff
@@ -15,6 +15,18 @@ module Moab
...
+    EXTRA_DIR_DETECTED = 2
+    CORRECT_DIR = 3
+    EMPTY = 4
+
+    RESPONSE_CODE_TO_MESSAGES = {
+      MISSING_DIR => "Missing directory: %{addl}",
+      EXTRA_DIR_DETECTED => "Unexpected item in path: %{addl}",
+      CORRECT_DIR=> "Correct items in path",
+      EMPTY => "No items in path"
+    }.freeze
```
[the remark this answers is not in the document]
Comment (SaravShah) [reply lib/moab/storage_repository.rb]: 👍

File: lib/moab/storage_repository.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (SaravShah) [reply lib/moab/storage_repository.rb]: Yeah so here i could either add a method that will check to see if it is named "data" or we will get a response msg from Under the version directory saying "Missing directory: data" and then a response msg from Under the (non-existent) data directory saying "No items in path"

It depends on how we want to do this... what do the other developers think?

File: lib/moab/storage_repository.rb [hunk 8]
```diff
@@ -139,6 +169,57 @@ def store_new_object_version(druid, bag_pathname )
       new_version
     end
 
-  end
+    private
+
+    def check_sub_dirs(sub_dir_count, sub_dirs, results, dir=nil)
```
[the remark this answers is not in the document]
Comment (SaravShah) [reply lib/moab/storage_repository.rb]: So something like ``` check_ver_dir ``` then ``` check_data_dir ``` , ``` check_manifest_dir ```, ```check_content_dir ```, and ``` check_metadata_dir ``` ??

[the remark this answers is not in the document]
Comment (ndushay) [reply lib/moab/storage_repository.rb]: in all seriousness?  yes.  Maybe they can call a common method with expected_sub_dirs and expected_files or something.

File: lib/moab/storage_repository.rb (hunk 2 above)
Comment (SaravShah) [reply lib/moab/storage_repository.rb]: Yeah i do like the idea of splitting it out into separate methods like: ``` check_ver_dir then check_data_dir , check_manifest_dir, check_content_dir, and check_metadata_dir ``` - i can issue a ticket for this. So for now are you saying get rid of the ``` sub_dir ``` method all together and implement this logic to either pick ``` EXPECTED_DATA_SUBDIRS = ["content", "metadata"].freeze or  EXPECTED_VERSION_SUBDIRS = ["data", "manifests"].freeze ``` within each method that was calling ``` sub_dir ``` ?

Comment (ndushay) [reply lib/moab/storage_repository.rb]: yes, except you can either use inline ["data", "manifest"]  or just use a class constant.

File: lib/moab/storage_repository.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (SaravShah) [reply lib/moab/storage_repository.rb]: yeah and then to your comment above then it would make everything more obvious to what each method is actually doing

File: lib/moab/storage_repository.rb [hunk 9]
```diff
@@ -115,19 +127,37 @@ def object_size(object_id, include_deposit=false)
       size
     end
 
+    # @param path [String] The identifier of the path to a digital object
+    # @return [Array] of hashes which contain messages for the caller
```
[the remark this answers is not in the document]
Comment (SaravShah) [reply lib/moab/storage_repository.rb]: yup, makes sense

File: lib/moab/storage_repository.rb (hunk 2 above)
Comment (SaravShah) [reply lib/moab/storage_repository.rb]: got it 👍

**Comment by SaravShah** [comment]
closed in favor of #72
